lib/dry/types/coercions.rb

```diff
@@ -8,18 +8,21 @@ def to_nil(input)
       end
 
       def to_date(input)
+        return input unless input.respond_to?(:to_str)
         Date.parse(input)
       rescue ArgumentError
         input
       end
 
       def to_date_time(input)
+        return input unless input.respond_to?(:to_str)
         DateTime.parse(input)
       rescue ArgumentError
         input
       end
 
       def to_time(input)
+        return input unless input.respond_to?(:to_str)
         Time.parse(input)
       rescue ArgumentError
         input
```

Comment:
I did that earlier, but unfortunately it doesn't work across all Ruby implementations.

Failed build for Rubinius:
https://travis-ci.org/dry-rb/dry-types/jobs/156124427

Failed Build for JRuby:
https://travis-ci.org/dry-rb/dry-types/jobs/156124428 

Switched to current method in this commit, with explanation:
https://github.com/dry-rb/dry-types/pull/138/commits/b47bbc167b5a1f7cfa702d94f6e7db7785aa89bf